pkg/test/framework/components/echo/echotest/run.go

```diff
@@ -15,31 +15,81 @@
 package echotest
 
 import (
+	"strings"
+
+	"istio.io/istio/pilot/pkg/networking/util"
 	"istio.io/istio/pkg/test/framework"
 	"istio.io/istio/pkg/test/framework/components/echo"
 )
 
 type (
-	perDeploymentTest func(t framework.TestContext, instances echo.Instances)
-	perInstanceTest   func(t framework.TestContext, src echo.Instance, dst echo.Instances)
+	perDeploymentTest  func(t framework.TestContext, instances echo.Instances)
+	perNDeploymentTest func(t framework.TestContext, deployments echo.Services)
+	perInstanceTest    func(t framework.TestContext, inst echo.Instance)
+
+	oneToOneTest func(t framework.TestContext, src echo.Instance, dst echo.Instances)
+	oneToNTest   func(t framework.TestContext, src echo.Instance, dsts echo.Services)
 )
 
-// Run will generate nested subtests for every instance in every deployment. The subtests will be nested including
-// the source service, source cluster and destination deployment. Example: `a/to_b/from_cluster-0`
-func (t *T) Run(testFn perInstanceTest) {
+// Run will generate and run one subtest to send traffic between each combination
+// of source instance to target deployment.
+//
+// For example, in a test named `a/to_b/from_cluster-0`,
+// `a` is the source deployment, `b` is the destination deployment and
+// `cluster-0` marks which instance of the source deployment.
+//
+// We use a combination of deployment name and cluster name to identify
+// a particular source instance, as there should typically be one instance
+// per cluster for any deployment. However we do not identify a destination
+// cluster, as we expect most tests will cause load-balancing across all possible
+// clusters.
+func (t *T) Run(testFn oneToOneTest) {
 	t.fromEachDeployment(t.rootCtx, func(ctx framework.TestContext, srcInstances echo.Instances) {
 		t.setup(ctx, srcInstances)
 		t.toEachDeployment(ctx, func(ctx framework.TestContext, dstInstances echo.Instances) {
-			t.setupPair(ctx, srcInstances, dstInstances)
-			t.fromEachCluster(ctx, srcInstances, dstInstances, testFn)
+			t.setupPair(ctx, srcInstances, echo.Services{dstInstances})
+			t.fromEachCluster(ctx, srcInstances, func(ctx framework.TestContext, src echo.Instance) {
+				filteredDst := t.applyCombinationFilters(src, dstInstances)
+				if len(filteredDst) == 0 {
+					// this only happens due to conditional filters and when an entire deployment is filtered we should be noisy
+					ctx.Skipf("cases from %s in %s with %s as destination are removed by filters ",
+						src.Config().Service, src.Config().Cluster.StableName(), dstInstances[0].Config().Service)
+				}
+				testFn(ctx, src, filteredDst)
+			})
+		})
+	})
+}
+
+// RunToN will generate nested subtests for every instance in every deployment subsets of the full set of deployments,
```

Comment:
This description is a bit hard to sift through.

IIUC this function sends N requests from every echo instance to a destination service (fqdn). And the main use case for this is where we expect some percentage of traffic to go to a certain subset of backends (e.g. traffic shifting) ... hence the N value. Is that right?

Reply:
From every instance (a deployment within a cluster) to N services (N deployments across all clusters).